### ISSUE TITLE
local resolver causes fetching state in query

### ISSUE THREAD

**Issue by louisholley** [FormidableLabs/urql#2097]
urql version & exchanges: 2.0.5, default exchanges + graphcache cacheExchange

**Steps to reproduce**

I've got a simple useQuery returning a list of jobs:
```
const [{ data, fetching, error }, reexecuteQuery] = useQuery({
    query: JOBS_LIST_QUERY,
  });

if (fetching) return <Loading />

...
```

I was trying to update what is served by the cache for the Job.organisation field, but I noticed that this problem persists even returning the parent field directly. When adding a local resolver, navigating to a different page and back to the job list shows the fetching state again, even though urql devtools show the results being served from the cache. The resolver is setup like this:
```
...
dedupExchange,
cacheExchange({
      resolvers: {
        Job: {
          organisation: (parent) => parent.organisation,
        },
      },
 }),
fetchExchange
...
```

**Expected behavior**
results should be served from cache and not show fetching again when query is cached

**Actual behavior**
fetching becomes true and UI shows loading despite the results being served from the cache

**Comment by JoviDeCroock** [comment]
I guess we're missing some information here, is the `data` actually populated as I'd assume you're returning `undefined` by leveraging `(parent) => parent.organisation,` which would rightfully result in a `loading` state.

**Comment by louisholley** [comment]
@JoviDeCroock thanks for the super quick response! i just checked and data does become undefined when the local resolver is present, is this supposed to happen? does the cache not just serve the result after going through the local resolver immediately?

fwiw, if i remove the local resolver then the app works as expected (navigating to and from the component with the cached query does not show the fetching state)...am i misunderstanding something? can try to work on a repro if that makes no sense

**Comment by JoviDeCroock** [comment]
That makes a lot of sense and let me tell you why, the resolvers are a way of you saying "let me help you out cache" and there are several things you can use this for:

- formatting a date
- changing a currency

The transformation case is documented [here](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#transforming-records)

These are the altering cases, there are other cases where you want to for instance give your cache a hand by telling it that `Query.organization` can be retrieved from the `Query.organizations` list.

The helping point at links/records is documented [here](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#resolving-entities)

The bottom line is: unless you need to transform or you know something isn't yet present in cache (a link/...) you'll need to leverage resolvers, in all other cases the retrieval logic will do everything for you.

**Comment by louisholley** [comment]
@JoviDeCroock thanks again, man. so what i'm trying to do is - i have a nullable field `job.organisationAlias` which i want the cache to return for the `job.organisation.name` field, but to return `job.organisation.name` if the alias is null. so I've got this:

```
...
cacheExchange({
  resolvers: {
    Job: {
      organisation: (parent) => ({
        ...parent,
        name: parent.organisationAlias ?? parent.organisation.name,
      }),
    },
}),
...
```

so the fields I'm requesting in the jobs list are:
```
jobs {
  id
  organisationAlias
  organisation {
    id
    name
  }
}
```

is this sort of thing not possible with the transformation case you describe without data becoming undefined and fetching becoming true?

**Comment by louisholley** [comment]
@JoviDeCroock I've got it working like this:
```
resolvers: {
  Job: {
    organisation: (parent, _args, cache) => {
      const organisationLink = cache.resolve(parent, "organisation");
      const organisationId = cache.resolve(organisationLink, "id");
      const organisationName = cache.resolve(organisationLink, "name");
      return {
        __typename: "Organisation",
        id: organisationId,
        name: parent.organisationAlias ?? organisationName,
      };
    },
  },
```

but I'm not entirely sure why the way i was trying initially wasn't working...is it because the field is a cache link so you can't return it directly? it's confusing because doing `console.log(parent.organisation)` in the resolver showed all the raw fields, not a cache ID...

if you can shed any light on this i'd be extremely grateful!

**Comment by JoviDeCroock** [comment]
Hmm, I'm not entirely sure why that was returning the full object, it could be due to timing and that object being mutated 😅 in that case it can show up in the console as populated while it was not at the time of you using it. You're spot on about the links. In the resolver you'll return _or_ `Organization:id` a cache-key _or_ the full object which conforms to the selection-set you have for `organization`. Sorry you were facing issues here, if we can do anything to clarify this in documentation feel free to point me towards the problem ❤️

**Comment by louisholley** [comment]
it's more just the console confusing me haha, i completely understand now! but this line:
> In the resolver you'll return or Organization:id a cache-key or the full object which conforms to the selection-set you have for organization

clarifies a lot for me, maybe i missed it in the docs for local resolvers but would be great to be explicit about it otherwise! thanks so much for your help again @JoviDeCroock !! loving urql so far
